Services/WebDriver: Enable on Windows

ayeteadoe opened the pull request:
This is the final relatively trivial target to get building on Windows. I believe all the remaining after this is:
1. LibCore/FileWatcher
2. LibCore/TimeZoneWatcher
3. LibCore/ProcessStatistics
4. test262-runner
5. LibLine
6. LibWasm/WASI

For 1-2 I have some in progress impls of them that I may look to complete soon-ish as well as 3 given it should be pretty straight forward.

For 4-6 see recent discussion in `#ui-windows` about them, so I'm not planning on looking into any of them myself for now.

ADKaster commented:
> LibCore/FileWatcher

Wait, do we even need to though? Looking at a quick grep, we only use Core::FileWatcher inside TimezoneWatcherLinux. So unless we have to watch a timezone file on Windows, or some other feature that needs to watch files for changes, we can probably keep it as unimplemented (and remove the macOS version?).

ADKaster commented:
Yeah, we could probably even demote it to implementation details of the Timezone watcher on linux. That discussion probably belongs on a removal PR with @trflynn89 in the loop though.

ayeteadoe commented:
See [this comment](https://discord.com/channels/1247070541085671459/1306918361732616212/1410178909609267261) in ui-windows for full context as to why I went with the following approach, but here are the changes that allowed a Qt-only Windows TimeZoneWatcher implementation:

<details>

<summary>windows_timezone_watcher.patch</summary>

```patch
diff --git a/Libraries/LibCore/CMakeLists.txt b/Libraries/LibCore/CMakeLists.txt
index d376e71932..4272696ead 100644
--- a/Libraries/LibCore/CMakeLists.txt
+++ b/Libraries/LibCore/CMakeLists.txt
@@ -83,19 +83,16 @@ if (LINUX AND NOT EMSCRIPTEN)
     list(APPEND SOURCES
         FileWatcherLinux.cpp
         Platform/ProcessStatisticsLinux.cpp
-        TimeZoneWatcherLinux.cpp
     )
 elseif (APPLE AND NOT IOS)
     list(APPEND SOURCES
         FileWatcherMacOS.mm
         Platform/ProcessStatisticsMach.cpp
-        TimeZoneWatcherMacOS.mm
     )
 else()
     list(APPEND SOURCES
         FileWatcherUnimplemented.cpp
         Platform/ProcessStatisticsUnimplemented.cpp
-        TimeZoneWatcherUnimplemented.cpp
     )
 endif()
 
diff --git a/Libraries/LibWebView/Application.cpp b/Libraries/LibWebView/Application.cpp
index b47a88408e..cb461ea4dd 100644
--- a/Libraries/LibWebView/Application.cpp
+++ b/Libraries/LibWebView/Application.cpp
@@ -9,7 +9,6 @@
 #include <LibCore/Environment.h>
 #include <LibCore/StandardPaths.h>
 #include <LibCore/System.h>
-#include <LibCore/TimeZoneWatcher.h>
 #include <LibDevTools/DevToolsServer.h>
 #include <LibFileSystem/FileSystem.h>
 #include <LibImageDecoderClient/Client.h>
@@ -19,6 +18,7 @@
 #include <LibWebView/Database.h>
 #include <LibWebView/HeadlessWebView.h>
 #include <LibWebView/HelperProcess.h>
+#include <LibWebView/TimeZoneWatcher.h>
 #include <LibWebView/URL.h>
 #include <LibWebView/UserAgent.h>
 #include <LibWebView/Utilities.h>
@@ -348,17 +348,18 @@ ErrorOr<void> Application::launch_services()
 
     // No need to monitor the system time zone if the TZ environment variable is set, as it overrides system preferences.
     if (!Core::Environment::has("TZ"sv)) {
-        if (auto time_zone_watcher = Core::TimeZoneWatcher::create(); time_zone_watcher.is_error()) {
+        if (auto time_zone_watcher = TimeZoneWatcher::create(); time_zone_watcher.is_error()) {
             warnln("Unable to monitor system time zone: {}", time_zone_watcher.error());
         } else {
             m_time_zone_watcher = time_zone_watcher.release_value();
 
-            m_time_zone_watcher->on_time_zone_changed = []() {
+            m_time_zone_watcher->set_on_time_zone_changed([]() {
+                dbgln("System time zone changed");
                 WebContentClient::for_each_client([&](WebView::WebContentClient& client) {
                     client.async_system_time_zone_changed();
                     return IterationDecision::Continue;
                 });
-            };
+            });
         }
     }
 
@@ -617,6 +618,13 @@ void Application::refresh_tab_list()
     m_devtools->refresh_tab_list();
 }
 
+Optional<TimeZoneWatcher&> Application::time_zone_watcher()
+{
+    if (m_time_zone_watcher != nullptr)
+        return *m_time_zone_watcher;
+    return {};
+}
+
 Vector<DevTools::TabDescription> Application::tab_list() const
 {
     Vector<DevTools::TabDescription> tabs;
diff --git a/Libraries/LibWebView/Application.h b/Libraries/LibWebView/Application.h
index 96d1013d1c..3fe862567c 100644
--- a/Libraries/LibWebView/Application.h
+++ b/Libraries/LibWebView/Application.h
@@ -28,6 +28,7 @@
 namespace WebView {
 
 struct ApplicationSettingsObserver;
+class TimeZoneWatcher;
 
 class WEBVIEW_API Application : public DevTools::DevToolsDelegate {
     AK_MAKE_NONCOPYABLE(Application);
@@ -71,6 +72,8 @@ public:
     ErrorOr<DevtoolsState> toggle_devtools_enabled();
     void refresh_tab_list();
 
+    Optional<TimeZoneWatcher&> time_zone_watcher();
+
 protected:
     explicit Application(Optional<ByteString> ladybird_binary_path = {});
 
@@ -143,7 +146,7 @@ private:
     OwnPtr<CookieJar> m_cookie_jar;
     OwnPtr<StorageJar> m_storage_jar;
 
-    OwnPtr<Core::TimeZoneWatcher> m_time_zone_watcher;
+    OwnPtr<TimeZoneWatcher> m_time_zone_watcher;
 
     OwnPtr<Core::EventLoop> m_event_loop;
     OwnPtr<ProcessManager> m_process_manager;
diff --git a/Libraries/LibWebView/CMakeLists.txt b/Libraries/LibWebView/CMakeLists.txt
index 8b8a43cf0a..b2f493e57d 100644
--- a/Libraries/LibWebView/CMakeLists.txt
+++ b/Libraries/LibWebView/CMakeLists.txt
@@ -34,6 +34,14 @@ set(SOURCES
 
 if (APPLE)
     list(APPEND SOURCES MachPortServer.cpp)
+elseif (LINUX AND NOT EMSCRIPTEN)
+    list(APPEND SOURCES TimeZoneWatcher/TimeZoneWatcherLinux.cpp)
+elseif (APPLE AND NOT IOS)
+    list(APPEND SOURCES TimeZoneWatcher/TimeZoneWatcherMacOS.mm)
+elseif (WIN32 AND ENABLE_QT)
+    list(APPEND SOURCES TimeZoneWatcher/TimeZoneWatcherWindowsQt.cpp)
+else()
+    list(APPEND SOURCES TimeZoneWatcher/TimeZoneWatcherUnimplemented.cpp)
 endif()
 
 set(GENERATED_SOURCES ${CURRENT_LIB_GENERATED})
@@ -96,6 +104,8 @@ if (ENABLE_QT)
     if (WIN32)
         find_package(pthread REQUIRED)
         target_include_directories(LibWebViewQt PRIVATE $<BUILD_INTERFACE:${PTHREAD_INCLUDE_DIR}>)
+        # NOTE: Required for the Qt-only Windows TimeZoneWatcher implementation
+        target_link_libraries(LibWebView PUBLIC Qt::Core)
     endif()
     target_link_libraries(LibWebViewPlatform INTERFACE LibWebViewQt)
 elseif (APPLE)
diff --git a/Libraries/LibWebView/TimeZoneWatcher.h b/Libraries/LibWebView/TimeZoneWatcher.h
new file mode 100644
index 0000000000..7747b123d8
--- /dev/null
+++ b/Libraries/LibWebView/TimeZoneWatcher.h
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2024, Tim Flynn <trflynn89@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Error.h>
+#include <AK/Function.h>
+#include <AK/Noncopyable.h>
+#include <AK/NonnullOwnPtr.h>
+#include <LibWebView/Export.h>
+
+#if defined(AK_OS_WINDOWS)
+class QAbstractNativeEventFilter;
+#endif
+
+namespace WebView {
+
+class WEBVIEW_API TimeZoneWatcher {
+    AK_MAKE_NONCOPYABLE(TimeZoneWatcher);
+
+public:
+    static ErrorOr<NonnullOwnPtr<TimeZoneWatcher>> create();
+    virtual ~TimeZoneWatcher() = default;
+
+    virtual void set_on_time_zone_changed(Function<void()> on_time_zone_changed)
+    {
+        this->on_time_zone_changed = move(on_time_zone_changed);
+    }
+
+#if defined(AK_OS_WINDOWS)
+    virtual QAbstractNativeEventFilter& qt_native_event_filter() = 0;
+#endif
+
+protected:
+    TimeZoneWatcher() = default;
+
+    Function<void()> on_time_zone_changed;
+};
+
+}
diff --git a/Libraries/LibWebView/TimeZoneWatcher/TimeZoneWatcherLinux.cpp b/Libraries/LibWebView/TimeZoneWatcher/TimeZoneWatcherLinux.cpp
new file mode 100644
index 0000000000..9f6373cd42
--- /dev/null
+++ b/Libraries/LibWebView/TimeZoneWatcher/TimeZoneWatcherLinux.cpp
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2024, Tim Flynn <trflynn89@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Platform.h>
+#include <LibCore/FileWatcher.h>
+#include <LibCore/TimeZoneWatcher.h>
+
+#if !defined(AK_OS_LINUX)
+static_assert(false, "This file must only be used for Linux");
+#endif
+
+namespace WebView {
+
+static constexpr auto time_zone_files = Array {
+    "/etc/localtime"sv,
+    "/etc/timezone"sv,
+    "/etc/TZ"sv,
+};
+
+static constexpr auto time_zone_mask = FileWatcherEvent::Type::ContentModified
+    | FileWatcherEvent::Type::Deleted
+    | FileWatcherEvent::Type::DoNotFollowLink;
+
+class TimeZoneWatcherImpl final : public TimeZoneWatcher {
+public:
+    static ErrorOr<NonnullOwnPtr<TimeZoneWatcherImpl>> create()
+    {
+        auto file_watcher = TRY(FileWatcher::create());
+
+        for (auto time_zone_file : time_zone_files) {
+            if (auto result = file_watcher->add_watch(time_zone_file, time_zone_mask); !result.is_error())
+                break;
+        }
+
+        return adopt_own(*new TimeZoneWatcherImpl(move(file_watcher)));
+    }
+
+private:
+    explicit TimeZoneWatcherImpl(NonnullRefPtr<FileWatcher> file_watcher)
+        : m_file_watcher(move(file_watcher))
+    {
+        m_file_watcher->on_change = [this](Core::FileWatcherEvent const& event) {
+            if (on_time_zone_changed)
+                on_time_zone_changed();
+
+            if (has_flag(event.type, FileWatcherEvent::Type::Deleted))
+                (void)m_file_watcher->add_watch(event.event_path, time_zone_mask);
+        };
+    }
+
+    NonnullRefPtr<FileWatcher> m_file_watcher;
+};
+
+ErrorOr<NonnullOwnPtr<TimeZoneWatcher>> TimeZoneWatcher::create()
+{
+    return TimeZoneWatcherImpl::create();
+}
+
+}
diff --git a/Libraries/LibWebView/TimeZoneWatcher/TimeZoneWatcherMacOS.mm b/Libraries/LibWebView/TimeZoneWatcher/TimeZoneWatcherMacOS.mm
new file mode 100644
index 0000000000..37ed6aadfd
--- /dev/null
+++ b/Libraries/LibWebView/TimeZoneWatcher/TimeZoneWatcherMacOS.mm
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2024, Tim Flynn <trflynn89@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Platform.h>
+#include <LibCore/TimeZoneWatcher.h>
+
+#if !defined(AK_OS_MACOS)
+static_assert(false, "This file must only be used for macOS");
+#endif
+
+#include <CoreFoundation/CoreFoundation.h>
+
+namespace WebView {
+
+class TimeZoneWatcherImpl final : public TimeZoneWatcher {
+public:
+    static ErrorOr<NonnullOwnPtr<TimeZoneWatcherImpl>> create()
+    {
+        return adopt_own(*new TimeZoneWatcherImpl());
+    }
+
+    virtual ~TimeZoneWatcherImpl() override
+    {
+        CFNotificationCenterRemoveObserver(
+            CFNotificationCenterGetLocalCenter(),
+            this,
+            kCFTimeZoneSystemTimeZoneDidChangeNotification,
+            nullptr);
+    }
+
+private:
+    explicit TimeZoneWatcherImpl()
+    {
+        CFNotificationCenterAddObserver(
+            CFNotificationCenterGetLocalCenter(),
+            this,
+            time_zone_changed,
+            kCFTimeZoneSystemTimeZoneDidChangeNotification,
+            nullptr,
+            CFNotificationSuspensionBehaviorDeliverImmediately);
+    }
+
+    static void time_zone_changed(CFNotificationCenterRef, void* observer, CFStringRef, void const*, CFDictionaryRef)
+    {
+        auto const& time_zone_watcher = *reinterpret_cast<TimeZoneWatcherImpl*>(observer);
+
+        if (time_zone_watcher.on_time_zone_changed)
+            time_zone_watcher.on_time_zone_changed();
+    }
+};
+
+ErrorOr<NonnullOwnPtr<TimeZoneWatcher>> TimeZoneWatcher::create()
+{
+    return TimeZoneWatcherImpl::create();
+}
+
+}
diff --git a/Libraries/LibWebView/TimeZoneWatcher/TimeZoneWatcherUnimplemented.cpp b/Libraries/LibWebView/TimeZoneWatcher/TimeZoneWatcherUnimplemented.cpp
new file mode 100644
index 0000000000..72be06963e
--- /dev/null
+++ b/Libraries/LibWebView/TimeZoneWatcher/TimeZoneWatcherUnimplemented.cpp
@@ -0,0 +1,16 @@
+/*
+ * Copyright (c) 2024, Tim Flynn <trflynn89@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCore/TimeZoneWatcher.h>
+
+namespace WebView {
+
+ErrorOr<NonnullOwnPtr<TimeZoneWatcher>> TimeZoneWatcher::create()
+{
+    return Error::from_errno(ENOTSUP);
+}
+
+}
diff --git a/Libraries/LibWebView/TimeZoneWatcher/TimeZoneWatcherWindowsQt.cpp b/Libraries/LibWebView/TimeZoneWatcher/TimeZoneWatcherWindowsQt.cpp
new file mode 100644
index 0000000000..c4179a8902
--- /dev/null
+++ b/Libraries/LibWebView/TimeZoneWatcher/TimeZoneWatcherWindowsQt.cpp
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2025, ayeteadoe <ayeteadoe@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Platform.h>
+#include <AK/Windows.h>
+#include <LibWebView/TimeZoneWatcher.h>
+
+#include <QAbstractNativeEventFilter>
+#include <QByteArray>
+
+#if !defined(AK_OS_WINDOWS)
+static_assert(false, "This file must only be used for Windows");
+#endif
+
+namespace WebView {
+
+class TimeChangeEventFilter : public QAbstractNativeEventFilter {
+public:
+    bool nativeEventFilter(QByteArray const& event_type, void* message, qintptr*) override
+    {
+        if (event_type == QByteArrayLiteral("windows_generic_MSG")) {
+            auto msg = static_cast<MSG*>(message);
+            if (msg->message == WM_TIMECHANGE && on_time_zone_changed) {
+                on_time_zone_changed();
+            }
+        }
+        return false;
+    }
+
+    Function<void()> on_time_zone_changed;
+};
+
+class TimeZoneWatcherImpl final : public TimeZoneWatcher {
+public:
+    static ErrorOr<NonnullOwnPtr<TimeZoneWatcherImpl>> create()
+    {
+        auto watcher = adopt_own(*new TimeZoneWatcherImpl());
+        TRY(watcher->initialize());
+        return watcher;
+    }
+
+    QAbstractNativeEventFilter& qt_native_event_filter() override
+    {
+        return m_time_change_event_filter;
+    }
+
+protected:
+    void set_on_time_zone_changed(Function<void()> on_time_zone_changed) override
+    {
+        m_time_change_event_filter.on_time_zone_changed = move(on_time_zone_changed);
+    }
+
+private:
+    TimeZoneWatcherImpl() = default;
+
+    ErrorOr<void> initialize()
+    {
+        return {};
+    }
+
+    TimeChangeEventFilter m_time_change_event_filter;
+};
+
+ErrorOr<NonnullOwnPtr<TimeZoneWatcher>> TimeZoneWatcher::create()
+{
+    return TimeZoneWatcherImpl::create();
+}
+
+}
diff --git a/UI/Qt/Application.cpp b/UI/Qt/Application.cpp
index 19d498901d..52f92c6c0c 100644
--- a/UI/Qt/Application.cpp
+++ b/UI/Qt/Application.cpp
@@ -14,6 +14,12 @@
 #include <QFileDialog>
 #include <QFileOpenEvent>
 
+#if defined(AK_OS_WINDOWS)
+#    include <AK/Windows.h>
+#    include <LibWebView/TimeZoneWatcher.h>
+#    include <QAbstractNativeEventFilter>
+#endif
+
 namespace Ladybird {
 
 class LadybirdQApplication : public QApplication {
@@ -27,6 +33,16 @@ public:
     {
         auto& application = static_cast<Application&>(WebView::Application::the());
 
+#if defined(AK_OS_WINDOWS)
+        static bool time_zone_watcher_installed = false;
+        auto time_zone_watcher = application.time_zone_watcher();
+        if (!time_zone_watcher_installed && time_zone_watcher.has_value()) {
+            QAbstractNativeEventFilter& event_filter = time_zone_watcher.value().qt_native_event_filter();
+            installNativeEventFilter(&event_filter);
+            time_zone_watcher_installed = true;
+        }
+#endif
+
         switch (event->type()) {
         case QEvent::FileOpen: {
             if (!application.on_open_file)
```

</details>

Any thoughts on this approach? Haven't put it up as a commit anywhere yet given I needed the current "Launch Ladybird on Windows" draft to test it. Unless Windows gets a non-qt based UI framework at some point in the future from this initial investigation I don't believe there will be another way to support this. So we could just leave it unimplemented for now too.
